torch/csrc/generic/Storage.cpp

```diff
@@ -31,14 +32,46 @@ static PyObject * THPStorage_(pynew)(PyTypeObject *type, PyObject *args, PyObjec
   THPStoragePtr self((THPStorage *)type->tp_alloc(type, 0));
   THPUtils_assert(self, "failed to allocate a " THPStorageStr " object");
   c10::Allocator* allocator = nullptr;
+  at::OptionalDeviceGuard device_guard;
 
   // Internally we allow constructing with a keywoard only argument cdata
   if (kwargs != nullptr) {
     PyObject *allocator_ptr = PyDict_GetItemString(kwargs, "allocator");
+    PyObject *device_ptr = PyDict_GetItemString(kwargs, "device");
+
+    TORCH_CHECK(!allocator_ptr || !device_ptr,
+        THPStorageStr, "(): cannot provide both 'allocator' and 'device' kwargs");
+
     if (allocator_ptr) {
       THPUtils_assert(THPUtils_checkLong(allocator_ptr), "invalid allocator");
       allocator = static_cast<c10::Allocator*>(PyLong_AsVoidPtr(allocator_ptr));
```

Comment:
Hmm, I just realized this seems kinda unsafe, doesn't it? The caller can provide any int64_t to the allocator arg and execute arbitrary code at that address in memory

Reply:
I think (but not 100% sure) the only reason this `allocator` kwarg exists is to implement `_StorageBase.pin_memory`, and I think that could be done in a safer way

Reply:
But I guess it's best to fix this later on